tfjs-backend-webgpu/src/kernels/matmul_packed_vec4_webgpu.ts

```diff
@@ -109,6 +109,58 @@ export function makeMatMulPackedVec4Source(workPerThread: number[]): string {
   `;
 }
 
+export function makeMatMulVectorVec4Source(): string {
+  return `
+    vec4 mm_readA(int row, int col);
+    vec4 mm_readB(int row, int col);
+    void mm_write(int row, int col, vec4 value);
+
+    const int TileSize = int(gl_WorkGroupSize.x) * 4;
+
+    shared vec4 mm_Asub[TileSize / 4];
+
+    void mm_matMul(int dimAOuter, int dimInner, int dimBOuter) {
+      int tileCol = int(gl_LocalInvocationID.x);
+      int globalCol = int(gl_GlobalInvocationID.x);
+      int globalRow = int(gl_GlobalInvocationID.y);
+
+      int numTiles = (dimInner - 1) / TileSize + 1;
+
+      vec4 acc = vec4(0.0);
+      vec4 ACached;
+
+      // Loop over shared dimension.
+      for (int t = 0; t < numTiles; t++) {
+        // Load one tile of A into local memory.
+        int colA = t * TileSize / 4 + tileCol;
+        mm_Asub[tileCol] = mm_readA(globalRow, colA);
+        barrier();
+
+        // Compute acc values for a single thread.
+        for (int k = 0; k < TileSize / 4; k++) {
+          int rowB = t * TileSize + k * 4;
+          vec4 BCached0 = mm_readB(rowB, globalCol);
+          vec4 BCached1 = mm_readB(rowB + 1, globalCol);
+          vec4 BCached2 = mm_readB(rowB + 2, globalCol);
+          vec4 BCached3 = mm_readB(rowB + 3, globalCol);
+
+          ACached = mm_Asub[k];
```

Comment:
Does this need to be declared outside the loop? Seems like it could be declared right here.

Reply:
Done